### DISCUSSION TITLE
I wonder how to use NeighborLoader correctly?

### DISCUSSION THREAD

**Discussion by PolarisRisingWar** [pyg-team/pytorch_geometric#3409] [Q&A]
I haven't learnt much about older version of NeighborSampler, just directly learnt NeighborLoader which is recommended.
And I find that each batch is a PyG.data.Data format graph, which seems that I can directly put one batch in a normal model which take `x,edge_index` as input. So is that the correct method to apply NeighborLoader?
Besides, is factors of `num_neighbors` paramater means how many neighbors to sample in this hop? And all sampled nodes and sampled edges consist of the graph of the output batch? And the `directed` input parameter decides whether the batch data is a directed graph, the `transform` input parameter decides if the batch data will be transformed?
I wonder what the `replace` input parameter decides, if it's false, it won't sample the same neighbor twice or more? So if it's true, will the batch data has edge weight?

And I've seen the example code for NeighborSampler in https://github.com/pyg-team/pytorch_geometric/blob/master/examples/reddit.py. It seems that the output batch was as many hops as many bipartite graphs sampled. And the code seems more complex, in training time it should use bipartite graphs to train mini-batches, in testing time it should collect node embeddings in all mini-batches for all layers, am I correct? I haven't understood it totally.

I'm changing a code in older version used NeighborSampler to a newer version. If my understanding is correct, I think it's much easier for me to use NeighborLoader than NeighborSampler. I can use it just like implementing full-batch GNN.

**Comment by rusty1s** [comment]
> Besides, is factors of `num_neighbors` paramater means how many neighbors to sample in this hop? And all sampled nodes and sampled edges consist of the graph of the output batch? And the `directed` input parameter decides whether the batch data is a directed graph, the `transform` input parameter decides if the batch data will be transformed? I wonder what the `replace` input parameter decides, if it's false, it won't sample the same neighbor twice or more? So if it's true, will the batch data has edge weight?

Yes, all your understandings are correct. `replace=False` ensures that we sample without replacement.

> And I've seen the example code for NeighborSampler in https://github.com/pyg-team/pytorch_geometric/blob/master/examples/reddit.py. It seems that the output batch was as many hops as many bipartite graphs sampled. And the code seems more complex, in training time it should use bipartite graphs to train mini-batches, in testing time it should collect node embeddings in all mini-batches for all layers, am I correct? I haven't understood it totally.

We deprecated `NeighborSampler` as the underlying implementation is indeed a little bit complex, in particular, because your model code now depends on the sampler you use. The `NeighborLoader` changes that, and can thus be seamlessly swapped out for other loaders as well, `ClusterLoader` or `GraphSAINTSampler`.

**Reply by PolarisRisingWar** [reply]
Thank you for you answers! And I have another problem, I wonder how could I map node in output graph of `NeighborLoader` to node in original graph? Is it sorted as original graph if I use `shuffle=False`?
I'm trying to use `train_mask`, `val_mask` and `test_mask` when set `input_nodes=None` for all nodes loader when testing. And this function may also be helpful when I want to study a specific case.

**Reply by rusty1s** [reply]
You should be able to access global indices if you add an additional property to your `data`, e.g.:
```python
data.n_id = torch.arange(data.num_nodes)

loader = NeighborLoader(data, ...)

for batch in loader:
    print(batch.n_id)
```
